### ISSUE TITLE
Big performance regression when scanning for workdir changes in repo with submodules

### ISSUE THREAD

**Issue by jeroen** [libgit2/libgit2#5725]
We are using a [metarepo](https://github.com/r-universe/ropensci) that contain submodules for many other projects. Recently we started noticing bad performance after updating to a new distro with a newer version of libgit2.

The regression happens in calls that scan for workdir changes, such as `git_status_list_new()` and `git_index_add_all`.

Below is a small example that illustrates the problem. What we want is the same as `git status` in the command line, to see if any of the submodules has been updated. But even if none of the submodules are actually initialized, this takes 2 sec. The git command line takes less than 0.1 sec for the same thing.

```c
#include <git2.h>
#include <stdio.h>

int main(){
	git_libgit2_init();
	git_repository *repo = NULL;
	if(git_repository_open(&repo, "ropensci")){
		fprintf(stderr, "Initial clone...");
		git_clone(&repo, "https://github.com/r-universe/ropensci", "ropensci", NULL);
		fprintf(stderr, "OK\n");
	}
	git_status_options opts = GIT_STATUS_OPTIONS_INIT;
	opts.flags = GIT_STATUS_OPT_INCLUDE_UNTRACKED |
		GIT_STATUS_OPT_RENAMES_HEAD_TO_INDEX |
		GIT_STATUS_OPT_SORT_CASE_SENSITIVELY;
	git_status_list *list = NULL;
	fprintf(stderr, "Running git_status_list_new...");
	git_status_list_new(&list, repo, &opts);
	fprintf(stderr, "done!\n");
	size_t len = git_status_list_entrycount(list);
	fprintf(stderr, "Found %d changes!\n", (int) len);
	return 0;
}
```

**Comment by ethomson** [comment]
Do you know the prior version that was less slow?  Do you have an idea of what that perf looked like?

**Comment by jeroen** [comment]
Looking into my past GHA logs now... performance for this call has actually always been bad, but it got 2x slower somewhere in between libgit2 0.28.4 and 1.0.0.

I have no idea what it can possibly be doing that is taking so long, if none of these submodules are even initialized?

**Comment by lhchavez** [comment]
if you call `git_repository_submodule_cache_all(repo);` (needs a `#include <git2/sys/repository.h>`) just before `git_status_list_new()`, it no longer needs to parse the `.gitmodules` file for every single submodule, making it as fast (or faster?) as shelling out.

**Comment by jeroen** [comment]
>if you call git_repository_submodule_cache_all(repo); (needs a #include <git2/sys/repository.h>) just before git_status_list_new(), it no longer needs to parse the .gitmodules file for every single submodule, making it as fast (or faster?) as shelling out.

Thanks! What is the reason that `git_repository_submodule_cache_all()` isn't documented in https://libgit2.org/libgit2/#HEAD/group/submodule or part of the regular exports in git2.h? Might it disappear one day?

Do I explicitly need to call `git_repository_submodule_cache_clear` to prevent leaking, or does it get cleaned when the we call `git_repository_free` ?

**Comment by ethomson** [comment]
> Thanks, that indeed fixes the problem! What is the reason that git_repository_submodule_cache_all() isn't documented in https://libgit2.org/libgit2/#HEAD/group/submodule 

I think that there's a documentation generation bug here.  😢 

(I would think that they would be in https://libgit2.org/libgit2/#HEAD/group/repository, not the submodule group, but that's a detail; your question is a good one either way.)

> or part of the regular exports in git2.h?

I'd have to familiarize myself with the submodule code - which, like all things submodules, I've done my best to avoid as best as possible.  But things go into `sys` when they're lower level functions that we don't expect that most (or any) people need to call.

I don't think that we've stated this explicitly, but in my thinking, things in `sys` are probably a little more experimental and I feel less bad about API changes.

> Might it disappear one day?

Yes, but not in the 1.x release series.  Exposing a function that instructs libgit2 to "cache things" has a real smell about it, and it feels like we should be doing the caching responsibly and evicting from the cache responsibly.  Now, maybe this simply isn't possible to do in a performant way with several hundred submodules, and this is precisely why this function exists.  But if we can just deal with this ourselves then that would be my preference.

What that would look like is these two functions becoming noops in 1.x and marked as deprecated, then removed entirely in a future release.

**Comment by jeroen** [comment]
Thank you for the detailed information. 

In my bindings, all of the `git_repository *` objects are short-lived, essentially a new one is allocated for each user command, similar as with command line git. 

AFAICT, it is safe for my bindings to globally enable `git_repository_submodule_cache_all()` for all commands then, because the cache will never outlive the temporary `git_repository *` object anyway... right?

Hopefully things will be fixed in https://github.com/libgit2/libgit2/pull/5727 and I can condition the workaround on `libgit2 < 1.1.1` 😄

**Comment by ethomson** [comment]
> AFAICT, it is safe for my bindings to globally enable git_repository_submodule_cache_all() for all commands then, because the cache will never outlive the temporary git_repository * object anyway... right?

Yes - with the caveat that you may be paying a penalty to load that cache if you're not doing work on submodules.  This sounds like it might not be a scenario that you need to worry about, though.

> Hopefully things will be fixed in #5727 and I can condition the workaround on libgit2 < 1.1.1 😄

Yeah, I think that it will, and frankly, I would just run it always.  You won't double-cache with #5727 since it will use the repo's submodule cache if you've created one.  (If I'm reading the code right.)
